File: 49GroupAnagrams.cpp
#include<iostream>
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp; 
        for (auto s: strs) {
            string t = s;
            sort(t.begin(), t.end());
            mp[t].push_back(s);
        }

        vector<vector<string>> answer;
        for (auto i: mp) {
            answer.push_back(i.second);
        }
        return answer;
        
    }
};

int main () {
    Solution s;
    vector<string> v = {"eat","tea","tan","ate","nat","bat"};
    vector<vector<string>> answer = s.groupAnagrams(v);

    for (int i = 0; i < answer.size(); i++) {
        for (int j = 0; j < answer[i].size(); j++) {
            cout << answer[i][j] << " ";
        }
        cout << endl;
    }
}

File: 49GroupAnagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        myMap = defaultdict(list) # dictionary of type list

        for word in strs:
            count = [0] * 26

            for char in word:
                count[ord(char) - ord("a")] += 1
                
            myMap[tuple(count)].append(word)

        return myMap

        
if __name__ == "__main__":
    s = Solution()
    strs = ["eat","tea","tan","ate","nat","bat"]
    s.groupAnagrams(strs)

